packages/api-bindings/src/apollo/createPublicationTypePolicy.ts

```diff
@@ -43,6 +50,18 @@ export function createPublicationTypePolicy(): TypePolicy<R> {
       isOptimisticMirroredByMe(existing: boolean | undefined) {
         return existing ?? false;
       },
+
+      collectState(
+        existing: CollectState | undefined,
+        { readField }: { readField: (field: keyof R) => Readonly<unknown> | undefined },
```

Comment:
TypeScript is struggling the infer `readField` with the correct parameters here, I imagine because of the generic type used:
```typescript 
type R = Post | Comment | Mirror;
```
I've been able to define a type to capture what we know it could return but unfortunately having to use `as ...` on the lines below.

Any ideas?

Reply:
```typescript
// Typescript not happy here
referencePolicy(
  existing: Readonly<R['referencePolicy']> | undefined,
  { readField }: { readField: ReadFieldFunction },
) {
  if (existing) return existing;

  const module = readField<ReferenceModule>('referenceModule');

  return resolveReferenceModule(module ?? null);
},
```

Saw this in my other PR as well.

Reply:
If I understood correctly, I think it's a mix of 2 things:
- `readField` is poorly typed to start with (in Apollo). This because it deals with normalized cache object. When using `readField` sometimes you get a so-called reference (a glorified identifier for the internal cache object) or the raw data (if it's a literal or it's an object and the cache is instructed to not normalize object of the given type via they corresponding `keyFields`). Now there is no easy way for the a generic TS type to figure out that a field pointing to an object (or array) is related to a type for which the cache normalization is turned off.
- the `TypePolicy` type we use is a patched version of the Apollo one (see `TypePolicy.ts`). It is something we partially inherited from an old project more than a year ago. I have reason to think that it does not account for the behaviour explained above. In other words it's an approximation that works in most of the cases.


Do you recognize in this anything familiar with the error you experienced?

Reply:
Turns out that despite the `reafField` type although an approximation of the real thing... is enough for this use case:
<img width="497" alt="image" src="https://user-images.githubusercontent.com/3353250/215766434-7bf0841d-faf9-4a2f-aaf2-16fab4f1b490.png">

Just by relying on the factory function return type it's enough to make it work.

I will be submitting a PR to fix this in Paul's Reference Policy PR.

I used `Post` instead of `R` (i.e. `Post | Comment | Mirror`) just to overcome another change I had to make (i.e. Mirror should not surface reference module settings).